### PR TITLE
orb-16

### PR DESCRIPTION
**Pull request by erichulser**
* supporting virtualization of columns and collectors

### CODE REVIEW

File: orb/core/column.py [hunk 1]
```diff
@@ -67,8 +67,8 @@ def __init__(self,
         self.__flags = self.Flags.fromSet(flags) if isinstance(flags, set) else flags
         self.__default = default
         self.__defaultOrder = defaultOrder
-        self.__getter = getter
-        self.__setter = setter
+        self.__getterName = getterName
```
Comment (erichulser) [inline orb/core/column.py]: this is a string, so it should not be confused with an actual getter method

File: orb/core/column.py [hunk 2]
```diff
@@ -484,12 +490,3 @@ def fromJSON(cls, jdata):
             col = col_cls()
             col.loadJSON(jdata)
             return col
-
-
-class VirtualColumn(Column):
```
Comment (erichulser) [inline orb/core/column.py]: removing virtualization this way in favor of more specific approach:

    @orb.virtual(BooleanColumn)
    def isEnabled(self):
        return False

File: orb/core/connection_types/sql/postgres/statements/alter.py [hunk 3]
```diff
@@ -27,6 +27,10 @@ def __call__(self, model, add=None, remove=None, owner='postgres'):
         add_i18n = []
         add_standard = []
         for col in add or []:
+            # virtual columns do not exist in the database
+            if col.testFlag(col.Flags.Virtual):
```
Comment (erichulser) [inline orb/core/connection_types/sql/postgres/statements/alter.py]: making sure virtual columns and collectors are not added to the database

File: orb/core/metamodel.py [hunk 4]
```diff
@@ -152,6 +152,11 @@ def store_property(key, value):
             for key, value in attrs.items():
                 if store_property(key, value):
                     attrs.pop(key)
+                elif hasattr(value, '__orb__'):
```
Comment (erichulser) [inline orb/core/metamodel.py]: `orb.virtual` will register an `__orb__` property to a method that will be understandable to the schema that is generated.

File: orb/core/model.py [hunk 5]
```diff
@@ -434,8 +434,8 @@ def get(self, column, default=None, useMethod=True, **context):
             context.inflated = False
 
         # call the getter method fot this record if one exists
-        elif useMethod:
-            method = getattr(type(self), col.getter(), None)
+        if useMethod:
```
Comment (erichulser) [inline orb/core/model.py]: virtual columns _have_ to be accessed via the getter

File: tests/conftest.py [hunk 6]
```diff
@@ -121,6 +120,16 @@ class User(orb.Table):
         groups = orb.Pipe(through='GroupUser', from_='user', to='group')
         userGroups = orb.ReverseLookup(from_column='GroupUser.user')
 
+        @orb.virtual(orb.BooleanColumn)
```
Comment (erichulser) [inline tests/conftest.py]: example of virtual column

File: tests/conftest.py [hunk 7]
```diff
@@ -121,6 +120,16 @@ class User(orb.Table):
...
+        def hasGroups(self):
+            return len(self.groups()) != 0
+
+        @orb.virtual(orb.Collector)
```
Comment (erichulser) [inline tests/conftest.py]: example of virtual collector

File: tests/general/test_schema.py [hunk 8]
```diff
@@ -109,4 +131,15 @@ def test_schema_name(GroupUser):
...
+def test_schema_json_export(User):
+    json = User.schema().__json__()
+
+    assert 'columns' in json
+    assert 'collectors' in json
+    assert json['model'] == 'User'
+    assert json['dbname'] == 'users'
+
+    # ensure virtual objects exist
+    assert 'my_groups' in json['collectors']
```
Comment (erichulser) [inline tests/general/test_schema.py]: ensuring virtual collector is included in the exported JSON schema

File: tests/postgres/test_pg_api.py [hunk 9]
```diff
@@ -373,3 +373,11 @@ def test_pg_api_reference_hash_id(orb, Comment, Attachment):
...
 
     assert isinstance(attachment.get('comment_id'), str)
+
+@requires_pg
+def test_pg_expand_virtual(orb, GroupUser, User):
+    gu = GroupUser.select().first().user()
+    u = User.select(where=orb.Query('id') == gu, expand='my_groups.users,groups.users').first()
+    json = u.__json__()
+    assert len(json['groups']) == len(json['my_groups'])
+    assert len(json['groups'][0]['users']) == len(json['my_groups'][0]['users'])
```
Comment (erichulser) [inline tests/postgres/test_pg_api.py]: ensuring expanding through a virtual collector works

File: orb/core/column.py (hunk 2 above)
Comment (danielmurray) [inline orb/core/column.py]: were we using virtual columns before?

File: orb/core/connection_types/sql/postgres/statements/create.py [hunk 10]
```diff
@@ -24,6 +24,10 @@ def _createTable(self, model, owner, includeReferences):
             if not includeReferences and isinstance(col, orb.ReferenceColumn):
                 continue
 
+            # virtual flags do not exist in the database
+            elif col.testFlag(col.Flags.Virtual):
```
Comment (danielmurray) [inline orb/core/connection_types/sql/postgres/statements/create.py]: what about virtual inputs with setter methods?
... might run into constraint satisfaction problem...

Comment (erichulser) [inline orb/core/connection_types/sql/postgres/statements/create.py]: You can do it this way:

    class User(orb.Table):
         @orb.virtual(orb.BooleanColumn)
         def isActive(self):
              return True
          @isActive.setter()
          def setActive(self, state):
               pass

Comment (erichulser) [inline orb/core/connection_types/sql/postgres/statements/create.py]: however, they would still not exist in the database...if they exist in the db, they shouldn't be virtual.

**Pull request merged**